source/air/src/printer.rs

```diff
@@ -495,3 +495,182 @@ impl NodeWriter {
+            // x % y ->  x - y*tmp
+            let pp = Printer::new(false);
+            let key = pp.expr_to_node(expr);
+            let value = node_map.get(&key);
+            let t = match value {
+                Some(tmp_var) => tmp_var.to_string(),
+                None => {
+                    let tmp_new = format!("tmp_{}", tmp_idx.to_string());
+                    *tmp_idx += 1;
+                    node_map.insert(key, tmp_new.to_string());
+                    tmp_new
+                }
+            };
+            let s1 = expr_to_singular(lhs, tmp_idx, node_map);
+            let s2 = expr_to_singular(rhs, tmp_idx, node_map);
+
+            format!("(({}) - (({})*({})))", s1, s2, t)
+        }
+        ExprX::Binary(BinaryOp::Eq, lhs, rhs) => {
+            let s1 = expr_to_singular(lhs, tmp_idx, node_map);
+            let s2 = expr_to_singular(rhs, tmp_idx, node_map);
+            format!("({}) - ({})", s1, s2)
+        }
+        ExprX::Multi(op, exprs) => {
+            let mut ss = vec![];
+            let sop = match op {
+                MultiOp::Add => ") + (",
+                MultiOp::Sub => ") - (",
+                MultiOp::Mul => ") * (", // still reachable with constant multiplication
+                _ => panic!("unsupported singular multi op"),
+            };
+            for e in &**exprs {
+                ss.push(expr_to_singular(&e, tmp_idx, node_map));
+            }
+            format!("(({}))", ss.join(sop))
+        }
+        ExprX::Apply(fname, exprs) => {
+            if "Mul" == (**fname).as_str() {
+                return expr_to_singular(
+                    &Arc::new(ExprX::Multi(MultiOp::Mul, exprs.clone())),
+                    tmp_idx,
+                    node_map,
+                );
+            } else if "EucMod" == (**fname).as_str() {
+                if exprs.len() != 2 {
+                    panic!("internal error: singular mod translation");
+                }
+                return expr_to_singular(
+                    &Arc::new(ExprX::Binary(
+                        BinaryOp::EuclideanMod,
+                        exprs[0].clone(),
+                        exprs[1].clone(),
+                    )),
+                    tmp_idx,
+                    node_map,
+                );
+            } else {
+                // treat as uninterpreted functions
+                let pp = Printer::new(false);
+                let key = pp.expr_to_node(expr);
+                let value = node_map.get(&key);
+                match value {
+                    Some(tmp_var) => tmp_var.to_string(),
+                    None => {
+                        let tmp_new = format!("tmp_{}", tmp_idx.to_string());
+                        *tmp_idx += 1;
+                        node_map.insert(key, tmp_new.to_string());
+                        tmp_new
+                    }
+                }
+            }
+        }
+        _ => panic!("unsupported singular expression: {:?}", expr),
```

Comment:
Can we fail more gracefully here? Or at least print a span? That may be easier if this happens in `rust_verify`. (Unless this is a redundant check of one done in an upper layer.)

Reply:
I guess I didn't use vir::def::MUL, because air does not depend on vir. (If I move these to `rust_verify` this would be solved as well)